contrib/dhcp_agent/dhcp_agent.py

```diff
@@ -0,0 +1,399 @@
+        hash_url.update(url.encode())
+        return hash_url.hexdigest()
+
+    def get_last_modified(self, url):
+        h = self.get_url_hash(url)
+        try:
+            last = self._config[CONFIG_HTTP_LAST_MODIFIED + h]
+        except KeyError:
+            last = None
+        return last
+
+    def set_last_modified(self, url, value):
+        h = self.get_url_hash(url)
+        self._config[CONFIG_HTTP_LAST_MODIFIED + h] = value
+
+    def fetch_configuration(self, mode, request_params):
+        is_saved = False
+        dhcp_config = self._get_configuration(mode, request_params)
+        if dhcp_config:
+            is_saved = self._set_new_configuration(
+                dhcp_config, mode
+            )
+        return all([dhcp_config, is_saved])
+
+    def get_request_params(self, dcs, envs):
+        request_params = []
+        if envs:
+            request_params.extend([('env', env) for env in envs])
+        if dcs:
+            request_params.extend([('dc', dc) for dc in dcs])
+        assert request_params != []
+        return request_params
+
+    def _get_configuration(self, mode, params):
+        url = '{}://{}/dhcp/{}/?{}'.format(
+            self.proto,
+            self.host,
+            mode,
+            urlencode(params)
+        )
+        configuration = None
+
+        self.logger.info('Sending request to {}'.format(url))
+        try:
+            response = self.make_request(url)
+        except HTTPError as e:
+            if e.code != 304:
+                self.logger.error(
+                    'Server returned {} status code with message "{}"'.format(
+                        e.code, e.fp.read().decode()
+                    )
+                )
+            else:
+                self.logger.info(
+                    'Server return status 304 NOT MODIFIED. Nothing to do.'
+                )
+            return False
+        else:
+            configuration = response.read()
+            self.set_last_modified(url, response.headers.get('Last-Modified'))
+        return configuration
+
+    def _send_synch_confirmation(self):
+        """This method notify Ralph about restarting DHCP server with
+        new configuration.
+
+        Returns:
+            bool: True if server returns 200 status code, otherwise False
+        """
+        url = '{}://{}/dhcp/synch/'.format(
+            self.proto,
+            self.host,
+        )
+        try:
+            self.make_request(url, data={})
+        except HTTPError as e:
+            self.logger.error(
+                'Could not send confirmation to Ralph. '
+                'Server returned {} status code with message: {}'.format(
+                    e.code, e.fp.read().decode()
+                ),
+            )
+            return False
+        self.logger.info('Confirmation sent to {}.'.format(self.host))
+        return True
+
+    def _set_new_configuration(self, config, mode):
+        """The method writing (or printing) config file.
```

Comment:
One function should do one job. This function is a hybrid of 2 functions and 'mode' parameter which function to choose. It's wrong. Why not pass file description (which stdout is 1 class citizen) and write to it with write() every time?  Mode parameter coud be ommited then.